invisible_cities/io/mcinfo_io.py

```diff
@@ -173,20 +177,180 @@ def load_mchits(file_name: str,
+
+        # Adding event info
+        evt_part_df = extents[['last_particle', 'evt_number']]
+        evt_part_df.set_index('last_particle', inplace = True)
+        parts = parts.merge(evt_part_df         ,
+                            left_index  =   True,
+                            right_index =   True,
+                            how         = 'left')
+        parts.rename(columns={"evt_number": "event_id"}, inplace = True)
+        parts.event_id.fillna(method='bfill', inplace = True)
+        parts.event_id = parts.event_id.astype(int)
+
+        # Setting the indexes
+        parts.set_index(['event_id', 'particle_id'], inplace=True)
+
+    return parts
+
+
 def load_mcsensor_response(file_name: str,
                            event_range=(0, int(1e9))) -> Mapping[int, Mapping[int, Waveform]]:
 
     with tb.open_file(file_name, mode='r') as h5in:
         return read_mcsns_response(h5in, event_range)
 
 
+def get_sensor_binning(file_name : str) -> Tuple:
+    """
+    Looks in the configuration table of the
+    input file and extracts the binning used
+    for both types of sensitive detector.
+    """
+    config   = pd.read_hdf(file_name, 'MC/configuration')
+    bins     = config[config.param_key.str.contains('time_binning')]
+    pmt_bin  = bins.param_value[bins.param_key.str.contains('Pmt')].iloc[0]
+    pmt_bin  = float(pmt_bin.split()[0]) * units_dict[pmt_bin.split()[1]]
+    sipm_bin = bins.param_value[bins.param_key.str.contains('SiPM')].iloc[0]
+    sipm_bin = float(sipm_bin.split()[0]) * units_dict[sipm_bin.split()[1]]
+
+    return pmt_bin, sipm_bin
+
+
+def load_mcsensor_response_df(file_name : str,
+                              db_file   : str,
+                              run_no    : int) -> Tuple:
+    """
+    A reader for the MC sensor output based
+    on pandas DataFrames.
+
+    file_name: string
+               Name of the file to be read
+    db_file  : string
+               Name of the detector database to be accessed
+    run_no   : int
+               Run number for database access
+    """
+    pmt_ids = DB.DataPMT(db_file, run_no).SensorID
+
+    pmt_bin, sipm_bin = get_sensor_binning(file_name)
+
+    extents  = pd.read_hdf(file_name, 'MC/extents')
+
+    sns      = pd.read_hdf(file_name, 'MC/waveforms')
+    evt_sns  = extents[['last_sns_data', 'evt_number']]
+    evt_sns.set_index('last_sns_data', inplace = True)
+
+    sns = sns.merge(evt_sns             ,
+                    left_index  =   True,
+                    right_index =   True,
+                    how         = 'left')
+    sns.evt_number.fillna(method='bfill', inplace = True)
+
+    sns['times'] = sns[sns.sensor_id.isin(pmt_ids)].time_bin * pmt_bin
+    sns.times.fillna(sns.time_bin * sipm_bin, inplace = True)
+
+    sns.evt_number = sns.evt_number.astype(int)
+    sns.set_index(['evt_number', 'sensor_id', 'time_bin'], inplace = True)
+
+    return extents.evt_number.unique(), pmt_bin, sipm_bin, sns
```

Comment:
Why does this function return also the list of event numbers and the other two don't?

Reply:
More a usage case than anything. The current function here is used in the new detsim where it was useful to have this information. The others are only used in FANAL so far where the extra info wasn't necessary

Reply:
As physicist, I like symmetries ... although a user case, it is a user case.